File: 22FALL/345.reverse-vowels-of-a-string.py
#
# @lc app=leetcode id=345 lang=python3
#
# [345] Reverse Vowels of a String
#

# @lc code=start
class Solution:
    def reverseVowels(self, s: str) -> str:
        characters = [c for c in s]
        i, j = -1, len(s)
        vowels = 'aeiouAEIOU'
        while i < j:
            while i+1 < len(s):
                i += 1
                if s[i] in vowels:
                    break
            while j-1 >= 0:
                j -= 1
                if s[j] in vowels:
                    break
            if i < j:
                characters[i], characters[j] = characters[j], characters[i]
        return ''.join(characters)

# @lc code=end

File: VSCode_CPP/345.reverse-vowels-of-a-string.cpp
/*
 * @lc app=leetcode id=345 lang=cpp
 *
 * [345] Reverse Vowels of a String
 */

// @lc code=start
class Solution {
public:
    string reverseVowels(string s) {
        string vowels = "aeiouAEIOU";
        string ans = s;
        int start = 0, end = s.size()-1;
        while (start < end)
        {
            while (start < end && vowels.find(s[start]) == -1)
                start++;
            while (start < end && vowels.find(s[end]) == -1)
                end--;
            swap(ans[start], ans[end]);
            start++;
            end--;
        }
        return ans;
        

    }
};
// @lc code=end

